# Q1. Write a Python function that takes in two arrays of data and calculates the F-value for a variance ratio test. The function should return the F-value and the corresponding p-value for the test.

In [2]:
import numpy as np
import scipy.stats as stats

def variance_ratio_test(data1, data2):
    """
    Perform a variance ratio test and return the F-value and p-value.

    Args:
    data1 (array-like): First array of data.
    data2 (array-like): Second array of data.

    Returns:
    float: F-value of the variance ratio test.
    float: p-value associated with the F-value.
    """
    # Calculate the variance of the two datasets
    var1 = np.var(data1, ddof=1)  # ddof=1 for unbiased estimate of variance
    var2 = np.var(data2, ddof=1)
    
    # Ensure var1 is the larger variance
    if var1 < var2:
        var1, var2 = var2, var1
    
    # Calculate the F-statistic
    f_statistic = var1 / var2
    
    # Calculate the degrees of freedom
    df1 = len(data1) - 1
    df2 = len(data2) - 1
    
    # Calculate the p-value
    p_value = stats.f.cdf(f_statistic, df1, df2)
    
    return f_statistic, p_value

# Example usage
data1 = [10, 12, 15, 8, 11, 13]
data2 = [7, 8, 9, 6, 10, 11]
f_value, p_value = variance_ratio_test(data1, data2)
print(f'F-value: {f_value}')
print(f'p-value: {p_value}')


F-value: 1.6857142857142857
p-value: 0.7097274320328928


# Q2. Given a significance level of 0.05 and the degrees of freedom for the numerator and denominator of an F-distribution, write a Python function that returns the critical F-value for a two-tailed test.

In [3]:
import scipy.stats as stats

def critical_f_value(alpha, dfn, dfd):
    """
    Calculate the critical F-value for a two-tailed test.

    Args:
    alpha (float): Significance level (e.g., 0.05 for 5% significance).
    dfn (int): Degrees of freedom for the numerator.
    dfd (int): Degrees of freedom for the denominator.

    Returns:
    float: Critical F-value for the given significance level and degrees of freedom.
    """
    # Calculate the critical F-value using the percent point function (ppf) of the F-distribution
    # for the given alpha and degrees of freedom
    f_critical = stats.f.ppf(1 - alpha / 2, dfn, dfd)
    
    return f_critical

# Example usage
alpha = 0.05  # Significance level
dfn = 3      # Degrees of freedom for the numerator
dfd = 20     # Degrees of freedom for the denominator
critical_f = critical_f_value(alpha, dfn, dfd)
print(f'Critical F-value: {critical_f}')


Critical F-value: 3.8586986662732143


# Q3. Write a Python program that generates random samples from two normal distributions with known variances and uses an F-test to determine if the variances are equal. The program should output the F- value, degrees of freedom, and p-value for the test.

In [4]:
import numpy as np
import scipy.stats as stats

# Set the random seed for reproducibility
np.random.seed(0)

# Define the parameters
sample_size = 30
variance1 = 1.5
variance2 = 2.0

# Generate random samples from two normal distributions with known variances
sample1 = np.random.normal(loc=0, scale=np.sqrt(variance1), size=sample_size)
sample2 = np.random.normal(loc=0, scale=np.sqrt(variance2), size=sample_size)

# Perform an F-test to compare the variances
f_statistic = np.var(sample1, ddof=1) / np.var(sample2, ddof=1)
df1 = sample_size - 1
df2 = sample_size - 1
p_value = 2 * min(stats.f.cdf(f_statistic, df1, df2), 1 - stats.f.cdf(f_statistic, df1, df2))

# Output the results
print(f'F-statistic: {f_statistic}')
print(f'Degrees of freedom (numerator): {df1}')
print(f'Degrees of freedom (denominator): {df2}')
print(f'p-value: {p_value}')

# Interpret the results
alpha = 0.05  # Set your desired level of significance

if p_value < alpha:
    print("The variances are not equal.")
else:
    print("The variances are equal.")


F-statistic: 1.0863928559446847
Degrees of freedom (numerator): 29
Degrees of freedom (denominator): 29
p-value: 0.8249607865956787
The variances are equal.


# Q4.The variances of two populations are known to be 10 and 15. A sample of 12 observations is taken fromeach population. Conduct an F-test at the 5% significance level to determine if the variances are significantly different.

To conduct an F-test to determine if the variances of two populations are significantly different, you can use the following steps. The null hypothesis (H0) is that the variances are equal, and the alternative hypothesis (H1) is that the variances are not equal.

Given:
- Population 1 variance (σ₁²) = 10
- Population 2 variance (σ₂²) = 15
- Sample size (n) = 12 for each population
- Significance level (α) = 0.05

Here's how you can perform the F-test:

1. Set up the hypotheses:
   - Null hypothesis (H0): σ₁² = σ₂² (the variances are equal).
   - Alternative hypothesis (H1): σ₁² ≠ σ₂² (the variances are not equal).

2. Calculate the test statistic (F-statistic) using the formula:
   
   F = (larger sample variance) / (smaller sample variance)

   In this case, the larger population variance is 15, and the smaller population variance is 10, so:

   F = 15 / 10 = 1.5

3. Determine the degrees of freedom for the numerator and denominator:
   - Degrees of freedom for the numerator (df1) is 11 (sample size - 1).
   - Degrees of freedom for the denominator (df2) is 11 for both populations since the sample sizes are the same.

4. Find the critical F-value for a two-tailed test at the 5% significance level. You can use an F-table or an online calculator to find the critical value. For α = 0.05 and df1 = 11 and df2 = 11, the critical F-value is approximately 3.816.

5. Compare the calculated F-statistic to the critical F-value:
   - If F-statistic > critical F-value, reject the null hypothesis (H0).
   - If F-statistic ≤ critical F-value, fail to reject the null hypothesis (H0).

In this case, the calculated F-statistic (1.5) is less than the critical F-value (3.816), so you would fail to reject the null hypothesis. This means that at the 5% significance level, there is not enough evidence to conclude that the variances of the two populations are significantly different.

#  Q5. A manufacturer claims that the variance of the diameter of a certain product is 0.005. A sample of 25 products is taken, and the sample variance is found to be 0.006. Conduct an F-test at the 1% significance level to determine if the claim is justified.

To conduct an F-test to determine if the manufacturer's claim about the variance of the product's diameter is justified, you can follow these steps. The null hypothesis (H0) is that the population variance is equal to the claimed value (0.005), and the alternative hypothesis (H1) is that the population variance is not equal to the claimed value.

Given:
- Manufacturer's claim about the population variance (σ²) = 0.005
- Sample variance (s²) = 0.006
- Sample size (n) = 25
- Significance level (α) = 0.01 (1%)

Here's how you can perform the F-test:

1. Set up the hypotheses:
   - Null hypothesis (H0): σ² = 0.005 (the population variance is equal to the claimed value).
   - Alternative hypothesis (H1): σ² ≠ 0.005 (the population variance is not equal to the claimed value).

2. Calculate the test statistic (F-statistic) using the formula:

   F = (sample variance) / (claimed population variance)

   F = 0.006 / 0.005 = 1.2

3. Determine the degrees of freedom for the numerator and denominator:
   - Degrees of freedom for the numerator (df1) is n - 1 = 25 - 1 = 24.
   - Degrees of freedom for the denominator (df2) is not specified, but it's typically large. You can assume a high value like 100 for this example.

4. Find the critical F-value for a two-tailed test at the 1% significance level. You can use an F-table or an online calculator to find the critical value. For α = 0.01, df1 = 24, and df2 = 100, the critical F-value is approximately 2.87.

5. Compare the calculated F-statistic to the critical F-value:
   - If F-statistic > critical F-value, reject the null hypothesis (H0).
   - If F-statistic ≤ critical F-value, fail to reject the null hypothesis (H0).

In this case, the calculated F-statistic (1.2) is less than the critical F-value (2.87), so you would fail to reject the null hypothesis. This means that at the 1% significance level, there is not enough evidence to conclude that the population variance of the product's diameter is significantly different from the claimed value of 0.005.

# Q6. Write a Python function that takes in the degrees of freedom for the numerator and denominator of an F-distribution and calculates the mean and variance of the distribution. The function should return the mean and variance as a tuple.

In [5]:
def f_distribution_mean_and_variance(dfn, dfd):
    """
    Calculate the mean and variance of an F-distribution.

    Args:
    dfn (int): Degrees of freedom for the numerator.
    dfd (int): Degrees of freedom for the denominator.

    Returns:
    tuple: Mean and variance of the F-distribution.
    """
    # Calculate the mean of the F-distribution
    mean = dfd / (dfd - 2) if dfd > 2 else None

    # Calculate the variance of the F-distribution
    variance = (2 * (dfd ** 2) * (dfn + dfd - 2)) / (dfn * (dfd - 2) ** 2 * (dfd - 4)) if dfd > 4 else None

    return mean, variance

# Example usage
dfn = 5  # Degrees of freedom for the numerator
dfd = 10  # Degrees of freedom for the denominator
mean, variance = f_distribution_mean_and_variance(dfn, dfd)
print(f'Mean: {mean}')
print(f'Variance: {variance}')


Mean: 1.25
Variance: 1.3541666666666667


# Q7. A random sample of 10 measurements is taken from a normal population with unknown variance. The sample variance is found to be 25. Another random sample of 15 measurements is taken from another normal population with unknown variance, and the sample variance is found to be 20. Conduct an F-test at the 10% significance level to determine if the variances are significantly different.

To conduct an F-test to determine if the variances of two populations are significantly different, you can follow these steps. The null hypothesis (H0) is that the population variances are equal, and the alternative hypothesis (H1) is that the population variances are not equal.

Given:
- Sample variance for the first population (s1²) = 25
- Sample variance for the second population (s2²) = 20
- Sample size for the first population (n1) = 10
- Sample size for the second population (n2) = 15
- Significance level (α) = 0.10 (10%)

Here's how you can perform the F-test:

1. Set up the hypotheses:
   - Null hypothesis (H0): σ1² = σ2² (the population variances are equal).
   - Alternative hypothesis (H1): σ1² ≠ σ2² (the population variances are not equal).

2. Calculate the test statistic (F-statistic) using the formula:

   F = (larger sample variance) / (smaller sample variance)

   In this case, the larger sample variance is 25, and the smaller sample variance is 20, so:

   F = 25 / 20 = 1.25

3. Determine the degrees of freedom for the numerator and denominator:
   - Degrees of freedom for the numerator (dfn) is n1 - 1 = 10 - 1 = 9.
   - Degrees of freedom for the denominator (dfd) is n2 - 1 = 15 - 1 = 14.

4. Find the critical F-value for a two-tailed test at the 10% significance level. You can use an F-table or an online calculator to find the critical value. For α = 0.10, dfn = 9 and dfd = 14, the critical F-value is approximately 3.488.

5. Compare the calculated F-statistic to the critical F-value:
   - If F-statistic > critical F-value, reject the null hypothesis (H0).
   - If F-statistic ≤ critical F-value, fail to reject the null hypothesis (H0).

In this case, the calculated F-statistic (1.25) is less than the critical F-value (3.488), so you would fail to reject the null hypothesis. This means that at the 10% significance level, there is not enough evidence to conclude that the population variances of the two samples are significantly different.

# Q8. The following data represent the waiting times in minutes at two different restaurants on a Saturday night: Restaurant A: 24, 25, 28, 23, 22, 20, 27; Restaurant B: 31, 33, 35, 30, 32, 36. Conduct an F-test at the 5% significance level to determine if the variances are significantly different.

To conduct an F-test to determine if the variances of the waiting times at Restaurant A and Restaurant B are significantly different, you can follow these steps. The null hypothesis (H0) is that the population variances are equal, and the alternative hypothesis (H1) is that the population variances are not equal.

Given:
- Waiting times at Restaurant A: [24, 25, 28, 23, 22, 20, 27]
- Waiting times at Restaurant B: [31, 33, 35, 30, 32, 36]
- Significance level (α) = 0.05 (5%)

Here's how you can perform the F-test:

1. Set up the hypotheses:
   - Null hypothesis (H0): σA² = σB² (the population variances are equal).
   - Alternative hypothesis (H1): σA² ≠ σB² (the population variances are not equal).

2. Calculate the sample variances for each restaurant:
   - Sample variance for Restaurant A (sA²) = 7.81 (you can calculate it using the formula for sample variance).
   - Sample variance for Restaurant B (sB²) = 7.33 (you can calculate it using the formula for sample variance).

3. Determine the degrees of freedom for the numerator and denominator:
   - Degrees of freedom for the numerator (dfn) is 6 - 1 = 5 (for Restaurant A).
   - Degrees of freedom for the denominator (dfd) is 5 - 1 = 4 (for Restaurant B).

4. Calculate the F-statistic using the formula:

   F = (larger sample variance) / (smaller sample variance)

   In this case, the larger sample variance is 7.81 (from Restaurant A), and the smaller sample variance is 7.33 (from Restaurant B), so:

   F = 7.81 / 7.33 ≈ 1.065

5. Find the critical F-value for a two-tailed test at the 5% significance level. You can use an F-table or an online calculator to find the critical value. For α = 0.05, dfn = 5 and dfd = 4, the critical F-value is approximately 6.055.

6. Compare the calculated F-statistic to the critical F-value:
   - If F-statistic > critical F-value, reject the null hypothesis (H0).
   - If F-statistic ≤ critical F-value, fail to reject the null hypothesis (H0).

In this case, the calculated F-statistic (1.065) is less than the critical F-value (6.055), so you would fail to reject the null hypothesis. This means that at the 5% significance level, there is not enough evidence to conclude that the population variances of waiting times at Restaurant A and Restaurant B are significantly different.

# Q9. The following data represent the test scores of two groups of students: Group A: 80, 85, 90, 92, 87, 83; Group B: 75, 78, 82, 79, 81, 84. Conduct an F-test at the 1% significance level to determine if the variances are significantly different.

To conduct an F-test to determine if the variances of the test scores in Group A and Group B are significantly different, you can follow these steps. The null hypothesis (H0) is that the population variances are equal, and the alternative hypothesis (H1) is that the population variances are not equal.

Given:
- Test scores for Group A: [80, 85, 90, 92, 87, 83]
- Test scores for Group B: [75, 78, 82, 79, 81, 84]
- Significance level (α) = 0.01 (1%)

Here's how you can perform the F-test:

1. Set up the hypotheses:
   - Null hypothesis (H0): σA² = σB² (the population variances are equal).
   - Alternative hypothesis (H1): σA² ≠ σB² (the population variances are not equal).

2. Calculate the sample variances for each group:
   - Sample variance for Group A (sA²) ≈ 16.33 (you can calculate it using the formula for sample variance).
   - Sample variance for Group B (sB²) ≈ 5.33 (you can calculate it using the formula for sample variance).

3. Determine the degrees of freedom for the numerator and denominator:
   - Degrees of freedom for the numerator (dfn) is 5 (for Group A).
   - Degrees of freedom for the denominator (dfd) is 5 (for Group B).

4. Calculate the F-statistic using the formula:

   F = (larger sample variance) / (smaller sample variance)

   In this case, the larger sample variance is approximately 16.33 (from Group A), and the smaller sample variance is approximately 5.33 (from Group B), so:

   F ≈ 16.33 / 5.33 ≈ 3.06

5. Find the critical F-value for a two-tailed test at the 1% significance level. You can use an F-table or an online calculator to find the critical value. For α = 0.01, dfn = 5 and dfd = 5, the critical F-value is approximately 9.966.

6. Compare the calculated F-statistic to the critical F-value:
   - If F-statistic > critical F-value, reject the null hypothesis (H0).
   - If F-statistic ≤ critical F-value, fail to reject the null hypothesis (H0).

In this case, the calculated F-statistic (approximately 3.06) is less than the critical F-value (approximately 9.966), so you would fail to reject the null hypothesis. This means that at the 1% significance level, there is not enough evidence to conclude that the population variances of test scores in Group A and Group B are significantly different.